### ISSUE TITLE
Triggering a popup by using a class or custom JS code

### ISSUE THREAD

**Issue by pbarmak** [elementor/elementor#7077]
## Prerequisites

- [x] I have searched for similar features requests in both open and closed tickets and cannot find a duplicate.
- [x] The feature is still missing in the latest stable version of Elementor ( Elementor Pro. )

**What problem is your feature request going to solve? Please describe.**
We used to use a plugin called Popup Maker. With their solution, each popup can be set with one or more custom classes, which, when used in a link or triggered via Javascript, would trigger the pop up.  It would be great to have this feature with Elementor popups, to have a special class which can be used in theme templates or other plugin code to trigger the popup open.

**Describe the solution you'd like**
Allow a custom class or multiple classes so we can trigger a popup from within JS or via a custom link, where it can be triggered outside of a specific Elementor button or link (like in a template).

**Describe alternatives you've considered**
As I mentioned, we've used Popup Maker's plugin for a while, good example of how it can be done, but I'd love to consolidate everything to Elementor and reduce plugins.

**Additional context**

**Comment by shilo-ey** [comment]
I merged this issue with  #6876, which is on the same subject so we could deal with it more efficiently. 
Please feel free to continue the conversation there.

Thanks for your feedback!

**Comment by sunnyt7** [comment]
Being able to trigger by CSS class would give the popups a lot more flexibility.

This is very useful for being able to trigger the popup with an element that doesn't have the dynamic URL field available, such as elements not created by Elementor, text within posts, or 3rd party addons that don't have the dynamic option built into the URL field.

**Comment by bentalgad** [comment]
+1 This would be amazing!

**Comment by bentalgad** [comment]
This is not the same subject as #6876 at all. This is giving the users the ability to set a class which will trigger the pop up so we can put this class even in things that our out of elementor widgets scope. for example a formidable pro form or any other thing.

**Comment by sunnyt7** [comment]
> 
> 
> This is not the same subject as #6876 at all. This is giving the users the ability to set a class which will trigger the pop up so we can put this class even in things that our out of elementor widgets scope. for example a formidable pro form or any other thing.

This is possible as of the last Pro 2.5 release. I disagree with its UI, but in the popup if you go to the Advanced tab, you can have it set to trigger by CSS class.

**Comment by bentalgad** [comment]
@sunnyt7 can you upload a screenshot? i don't see that. i see the option to choose selector for scrolling to element, but not clicking element.

**Comment by taropaa** [comment]
@bentalgad Here is an example I am triggering via a menu option. Very risky since the ID of the menu item can change. 
![Screen Shot 2019-04-08 at 5 19 56 PM](https://user-images.githubusercontent.com/727908/55757690-96637880-5a22-11e9-9b22-a3448ec198b4.png)

**Comment by bentalgad** [comment]
Oh, got it @taropaa . thanks. that's a very weird decision to place it there and not in the conditions dialog. but happy to know it's possible :)

**Comment by bkjproductions** [comment]
I need to do this via Javascript. I have a series of links, each which has HTML "data" that I need to pass along to a popup.
`<a href="#" data-building="Power Station" class="building-map">Power Station</a>`
So when you click on an item, then it pops up the popup and within that another listener does something else within the popup.
```
	$('.building-map').on('click',function() {
		var my_building = $(this).data('building');
		alert('need to show: ' + my_building);
	});
```
So what I would like to know is how to trigger the popup simply via Javascript, NOT using the built-in class selector (which is nice, of course, but not for this situation, because I need to grab the data from the item clicked).

**Comment by setras** [comment]
mybe this plugin helps
[https://github.com/sukiwp/popup-trigger-url-for-elementor-pro](https://github.com/sukiwp/popup-trigger-url-for-elementor-pro)

**Comment by userabuser** [comment]
@bkjproductions (and for anyone else looking for a JS only solution) you can use:

```js
// accessible on window.elementorProFrontend
elementorProFrontend.modules.popup.showPopup( { id: 123 } );
```

In the example above `123` corresponds to the document ID (Post ID) of the popup which you can derive many ways, but within the DOM it's the ID of the containing element:

 `<div id="elementor-popup-modal-123">...</div>`

Known to work in version 2.8.* and of course with Elementor Pro.

`frontend.js` is a dependency (handle: `elementor-frontend`) so make sure you are executing this call after `frontend.js` is enqueued and printed.

**Comment by jackslayer1234** [comment]
> @bkjproductions (and for anyone else looking for a JS only solution) you can use:
> 
> ```js
> // accessible on window.elementorProFrontend
> elementorProFrontend.modules.popup.showPopup( { id: 123 } );
> ```
> 
> In the example above `123` corresponds to the document ID (Post ID) of the popup which you can derive many ways, but within the DOM it's the ID of the containing element:
> 
> `<div id="elementor-popup-modal-123">...</div>`
> 
> Known to work in version 2.8.* and of course with Elementor Pro.
> 
> `frontend.js` is a dependency (handle: `elementor-frontend`) so make sure you are executing this call after `frontend.js` is enqueued and printed.

Wow, this actually worked. thanks! By the way, is it possible to set conditions eg. popup appears once only?

For example, if I did the following, then each time a user clicks `billing_first_name`, the popup appears.

```
document.getElementById("billing_first_name").onclick = function(){
    elementorProFrontend.modules.popup.showPopup( { id: 26718 } );

}
```

**Comment by unostar** [comment]
This is short jQuery solution to open Popup from within code. Just place the correct Popup document id.
```
jQuery(window).on('elementor/frontend/init', function() {
  elementorFrontend.on( 'components:init', function() {
    elementorFrontend.documentsManager.documents[id].showModal();
  });
});
```

**Comment by acal** [comment]
Thanks, @taropaa That worked for me. I used a class selector instead since I use the trigger on more than one element in a page. Works great on both elementor widgets and custom html.

**Comment by TCattd** [comment]
As @unostar already mentioned, you can open a popup with this:
`elementorFrontend.documentsManager.documents[id].showModal();`

id is the data-elementor-id value for that popup.

Now, you can close a modal via JS too. You only need to call this other method:
`elementorFrontend.documentsManager.documents[id].getModal().hide();`

Again, replace id with the proper value for your popup.

This is useful so you can bind a button ID inside the popup, to close it programatically from outside Elementor.

Elementor really need a better API for developers. Or documentation for the one they already use actually.

**Comment by wahidmagdy** [comment]
@unostar 's answer with 
`elementorFrontend.documentsManager.documents[id].showModal();`
worked better for me than the other answer more commonly referenced 
`elementorProFrontend.modules.popup.showPopup( { id: 26718 } );`

for some unknown reason, sometimes the page would load and `elementorProFrontend.modules` would just be empty (even after waiting for the `elementor/frontend/init` event).

Calling the modal through elementorFrontend instead of elementorProFrontend has been a lot more stable for me. hopefully moving forward there is better developer documentation available

**Comment by leon2810** [comment]
do it still works for you?

I get an error :

`jquery-3.6.0.min.js?ver=3.6.0:2 Uncaught TypeError: this.getModal is not a function
    at _default.showModal (elements-handlers.min.js?ver=3.5.2:2)
    at Frontend.<anonymous> ((index):824)
    at Function.<anonymous> (frontend-modules.min.js?ver=3.5.3:2)
    at Function.each (jquery-3.6.0.min.js?ver=3.6.0:2)
    at Frontend.Module.trigger (frontend-modules.min.js?ver=3.5.3:2)
    at Frontend.e.each.n.<computed> [as trigger] (frontend-modules.min.js?ver=3.5.3:2)
    at Frontend.onDocumentLoaded (frontend.min.js?ver=3.5.3:2)
    at Frontend.initOnReadyComponents (frontend.min.js?ver=3.5.3:2)
    at Frontend.init (frontend.min.js?ver=3.5.3:2)
    at HTMLDocument.<anonymous> (frontend.min.js?ver=3.5.3:2)`

**Comment by edwardk1234** [comment]
It is not works

**Comment by leon2810** [comment]
apparently, it works for me only when I'm logged in to wp-admin

**Comment by itkampioen** [comment]
> apparently, it works for me only when I'm logged in to wp-admin

Having the same issue here. Apparently the JS API is different when logged in... Any fix for this?

**Comment by Chielob** [comment]
Both the options work in my case. However, the pop-up conditions are neglected. Is there any way on trigger the pop-up using JS with taking into account the set conditions in Elementor (e.g. show only 5 times)?

**Comment by bkjproductions** [comment]
@Chielob  I would say that if you are manually triggering a popup, you could NOT expect the Display Conditions such as show-only-5-times to be followed. In the manual-trigger situation you are asking for something specific that is not one of the conditions. Perhaps, though, if one of the conditions were "on exit intent" you could track down what that "event" was, and call that. An easy one might be "on click" -- so you could send a "click" to some element on the page, and maybe it would work. 

But in the use case of triggering a popup by clicking on a button or menu item, you would still want the popup to display regardless of conditions such as X sessions or X page views or previous URL, etc. Otherwise the button or menu item would appear not to work. So I think it would make sense not to allow any of the conditions to be considered, save the one "Include in entire site" (so that the necessary code for the popup exists on the page).

**Comment by thd-fox** [comment]
I have the same situation like @leon2810.

**Found a workaround for anonymous visitors:**

I'm lucky enough to distinguish display popup condition with url get param.
I have a url param of ?popup=thanks

In theme functions.php file I then past this code:
```
if ( isset( $_GET['popup'] ) && $_GET['popup'] == "thanks" ) {
     echo '<div class="trigger-popup"></div>';
}
```

I use trigger on scroll to element. As the above element is added in functions.php without any hook, it is already on scroll on page load.

If you have any other condition, you can achieve this from JS.
Maybe append div element anywhere in your current view and trigger scroll event.

**Comment by tomektomczuk** [comment]
I managed to trigger popup this way `add_action( 'wp_footer', 'spanish_popup', 10);
function spanish_popup(){

        $popup_id = '123179'; //your Popup ID.
        ElementorPro\Modules\Popup\Module::add_popup_to_location( $popup_id ); //insert the popup to the current page

        ?><script>
            jQuery( document ).ready( function() {
                jQuery( window ).on( 'elementor/frontend/init', function() {
                    elementorFrontend.on( 'components:init', function() {
                        if (jQuery("html[lang='es-ES']").length) {
                            elementorFrontend.documentsManager.documents[<?php echo $popup_id ;?>].showModal();
                        }
                    } );
                } );
            } );
        </script>;
        <?php
 }`

other solutions trigger Uncaught TypeError: this.getModal or showModal()

**Comment by woodydaniel** [comment]
@tomektomczuk I'm still getting errors using this method. any ideas?

**Comment by juanradark** [comment]
I think the problem for most people is that you have to add a Rule to insert the popup in the pages you want it to show and then:

`elementorFrontend.documentsManager.documents[2198].showModal();`

**Comment by woodydaniel** [comment]
@juanradark I have added the rule. It seems to work when I'm logged in as admin, but if i'm logged out (like a normal customer) it displays the uncaught type error showModal()

**Comment by oshevans** [comment]
As with many above, both code samples work when logged in as admin, but neither work when logged in as subscriber.
Popup has condition to show an all pages, but also tried add_popup_to_location - no difference.

`console.log(elementorFrontend.documentsManager.documents[id]);` shows an empty element when not logged in as admin.

Anyone resolve this?

**Comment by Zeeshan99** [comment]
In case, you have two popups and wanted to trigger the previous popup from the element inside the second popup. All you just need to insert HTML widget in the Popup and insert the below script: 

```
<script>
jQuery(document).ready(function(){
jQuery("#showMenuMenu").on("click", function(){

**//Please replace the "popupId" with the real ID e.g 9876.**
elementorFrontend.documentsManager.documents[popupId].getModal().hide();
setTimeout(function() {

**//Please replace the "popupId" with the real ID e.g 9878.**
elementorProFrontend.modules.popup.showPopup( { id: popupId } );
 }, 700);
});
});    
</script>

```
Please note that if you are using the Popup animations, your time must match the time in the above function.
For instance, here the time is **700** so the Popup animation's time must be **0.7**

**Comment by tailorhound** [comment]
I think what's creating a lot of confusion here is that it takes a while for elementorFrontend to initialise.  I'm doing this:

```
document.addEventListener('DOMContentLoaded', init);
function init(){
    setTimeout(showPopup, 10);
}
function showPopup(){
    elementorFrontend.documentsManager.documents[175].showModal();
}
```

and that extra 10ms is enough of a pause to give elementorFrontend time to initialise.  Interestingly, even 0ms seems to work - as the timeout is still introducing a slight processing delay.  However this:

```
document.addEventListener('DOMContentLoaded', init);
function init(){
    elementorFrontend.documentsManager.documents[175].showModal();
}

```
Gives me: Uncaught TypeError: Cannot read properties of undefined (reading 'documents')

It's a bit janky, I wish there was an event saying when we could use elementorFrontend, but this is better than nothing.

**Comment by nicholaszein** [comment]
This feature was released with Elementor Pro **v2.5**
